### DISCUSSION TITLE
Resource management

### DISCUSSION THREAD

**Discussion by cameron-martin** [koka-lang/koka#395] [Q&A]
Is there a pattern for resource management in koka, e.g. for files, sockets, etc? I see that the standard library avoids this entirely, for example `std/os/file` only has functions for reading and writing entire files.

**Comment by chtenb** [comment]
Perhaps the `finally` handler is already enough for most use cases:

```
fun test-finally() 
  with finally{ println("exiting..") } // dispose resources here
  println("entering..")
  throw("oops") + 42
```

If there were a need for disposing some resource, the dispose logic could be placed such a finally handler. This would be similar to how a language like C# does this through their `try/finally` blocks.

But I'm not sure if you could hook that into the type/effect system, such that undisposed resources would be flagged, and I haven't encountered a standard pattern for doing so. Interesting topic.
